server/matchmaker/map_pool.py

```diff
@@ -0,0 +1,41 @@
+import random
+from collections import Counter
+from typing import List
+
+from ..decorators import with_logger
+from ..types import MapDescription
+
+
+@with_logger
+class MapPool(object):
+    def __init__(self, name: str, maps: List[MapDescription] = []):
+        self.name = name
+        self.set_maps(maps)
+
+    def set_maps(self, maps: List[MapDescription]) -> None:
+        self.maps = {map_.id: map_ for map_ in maps}
+        self.map_ids = set(self.maps.keys())
+
+    def choose_map(self, played_map_ids: List[int] = []) -> MapDescription:
+        """
+        Select a random map who's id does not appear in `played_map_ids`. If
+        all map ids appear in the list, then pick one that appears the least
+        amount of times.
+        """
+        if not self.maps:
+            self._logger.critical(
+                "Trying to choose a map from an empty map pool: %s", self.name
+            )
+            raise RuntimeError(f"Map pool {self.name} not set!")
+
+        if set(played_map_ids) >= self.map_ids:
+            # Every map was played, find the least common one
+            map_id = Counter(played_map_ids).most_common()[-1][0]
+            return self.maps[map_id]
```

Comment:
Flagging this again, since my previous review was already outdated when I submitted it:
Is `map_id` guaranteed to be in `self.map_ids` or should  this be something like the following?
```python
map_id = Counter(i for i in played_map_ids if i in self.map_ids)
```